### ISSUE TITLE
"Bootstrapping" 32-bit => 64-bit RaspiOS Lite / DietPi

### ISSUE THREAD

**Issue by xanoni** [keks24/raspberry-pi-luks#12]
**TL;DR:**

- Rationale for choosing this route: minimize downtime of an existing 32-bit RPi system while migrating it to 64-bit (in the absence of another TRUSTED device to do this work on)
- Didn't expect any issues, but unfortunately ran into problems with `qemu` and `binfmt`, which  can be solved easily once they are understood (i.e., going through the process again based on below instructions would be much faster than what it took me to figure it out)
- If you don't have any production services on your 32-bit Pi or if you have another trusted machine where you can prepare the image, you probably still don't want to choose this route ... working through the guide on a 32-bit machine just introduces too many random problems, as I've learned (see below)

**Orignal story:**

I'm currently trying to transform a Pi that runs 32-bit RaspberryPi OS Light into one that runs 64-bit DietPi. Given I never used DietPi before, I decided to first try upgrading to the "official" 64-bit RaspiOS Lite image. Specifically, I'm using [the latest version from May (v5.10.52)](https://downloads.raspberrypi.org/raspios_lite_arm64/images/raspios_lite_arm64-2021-05-28/).

This project is still work-in-progress, but want to document my progress here so maybe I can help with getting the guide ready for the upcoming stable 64-bit release. After the "proof-of-concept" succeeds, I'll see what the easiest way to turn it into a DietPi would be (which is based on the same repositories/packages and uses the same kernel). There's some overhead, but its more important to me to eliminate unknowns, given how many days I spent last time finding/fixing random details.

**QEMU for chroot:**

As we're trying to chroot into a 64-bit ARM image from a 32-bit ARM machine, we have to use `qemu-aarch64-static` instead of `qemu-arm-static`:

```bash
# the below is entered OUTSIDE the chroot
# FYI, I have other stuff in /mnt/ already, so I use /mnt/croot/ instead (= "cryptroot")
apt install qemu-user-static
export CHROOT="/mnt/croot"
cp /usr/bin/qemu-aarch64-static "${CHROOT}/usr/bin/"
chroot "${CHROOT}" qemu-aarch64-static /bin/bash
```

Unfortunately, there seems to be some problem with `binfmt` which means that 64-bit files are not recognized in the `chroot`. Specifically, `qemu-aarch64` was not showing up in `update-binfmts --display` outside the `chroot`. Thus, any commands entered after launching the chroot failed. In theory, this can be fixed via:

```bash
# the below is entered OUTSIDE the chroot
update-binfmts --enable qemu-aarch64
```

Unfortunately that didn't work, so I first had to manually define `qemu-aarch64` before enabling it:

```bash
# the below is entered OUTSIDE the chroot
export magic='\x7f\x45\x4c\x46\x02\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\xb7\x00'
export mask='\xff\xff\xff\xff\xff\xff\xff\x00\xff\xff\xff\xff\xff\xff\xff\xff\xfe\xff\xff\xff'
update-binfmts --package qemu-user-static --install qemu-aarch64 /usr/bin/qemu-aarch64-static --magic "$magic" --mask "$mask" --offset 0 --credential yes --fix-binary yes
```

The 64-bit kernel uses the versioning format `5.10.52-v8+`, so I changed that as follows wherever you had `5.10.17+` in the guide (I don't want to keep building the `-v7*` stuff, given how long it takes on the Pi, so I don't just use `5.10.52+`):

```bash
# the below is entered INSIDE the chroot
export KERNEL_TARGET_VERSION="5.10.52-v8+"
mkdir --parents "/etc/kernel/{postinst.d,postrm.d}/${KERNEL_TARGET_VERSION}/" "/etc/kernel/preinst.d/"
ln --symbolic --verbose "/usr/local/share/kernel/postinst.d/01-rpi-initramfs-tools" "/etc/kernel/postinst.d/${KERNEL_VERSION}/"
ln --symbolic --verbose "/usr/local/share/kernel/postrm.d/01-rpi-initramfs-tools" "/etc/kernel/postrm.d/${KERNEL_TARGET_VERSION}/"
ln --symbolic --verbose "/usr/local/share/kernel/preinst.d/01-rpi-initramfs-tools" "/etc/kernel/preinst.d/"
```

Unfortunately, the next step cost me several hours because I was long overdue to go to bed, but didn't want to give up... which made it worse in the end, but whatever:

The creation of the `initramfs` always mysteriously failed towards the very end of the process—which takes LONG if running via `qemu` on a 32-bit Pi. I tried all 3 options that you list in the guide (`--reinstall` kernel, `update-initramfs`, ...), but no difference as the error was in `mkinitramfs` itself.

Ultimately, I debugged the scripts step-by-step and it turned out that choosing another compression algorithm fixes it ... when the default `gzip` is selected, then the scripts actually run `pigz`, which is a variation optimized for parallel execution. However, as we noticed it fails for mysterious reasons. Thus, we swap it out for `bzip2`.

```bash
# the below is entered INSIDE the chroot
vi /etc/initramfs-tools/initramfs.conf # change "COMPRESS=gzip" to "COMPRESS=bzip2"
apt-get uninstall raspberrypi-kernel # doing this because --reinstall sometimes failed
apt-get install raspberrypi-kernel
```

I still have one detail left to be tackled... the machine doesn't boot, yet, and I don't have a screen to debug it. I will update this ticket as I make progress.

**Comment by xanoni** [comment]
```
Please unlock disk croot:
cryptsetup: croot set up successfully
Connection to 192.168.23.42 closed.
```

I think that's all, the reason why it didn't work yesterday is because I am using a GPT disklabel instead of the standard MBR/DOS label and then I also tweaked the boot partition a bit too much. I was able to keep the GPT label, but went back to the standard boot partition structure as found in the *.img file.

`mkinitramfs` with `gzip` / `pigz` still doesn't work, but I successfully tested `bzip2` and `lzma`. I assume now that I'm booted into 64-bit and no longer using `qemu-aarch64-static`, the standard `gzip` / `pigz` will work, too. Will update this once I've tested it.

**Comment by xanoni** [comment]
Added a TL;DR in the original post above given that my motivation for choosing this route was not obvious.

I'm likely in non-standard circumstances to be choosing this route. However, I still hope it helps at least 1 other person.

**Comment by keks24** [comment]
> The creation of the initramfs always mysteriously failed towards the very end of the process—which takes LONG if running via qemu on a 32-bit Pi. I tried all 3 options that you list in the guide (--reinstall kernel, update-initramfs, ...), but no difference as the error was in mkinitramfs itself.

Maybe `pigz` is doing something weird here.

Check, if the output of the `file` command looks something like this:
```bash
$ file /boot/initramfs.cpio.gz
/boot/initramfs.cpio.gz: gzip compressed data, last modified: Sat Sep  4 08:15:56 2021, from Unix, original size 25040896
```

On my Gentoo system it looks like this:
```bash
$ file /usr/src/custom-initramfs-5.10.61-gentoo.cpio.gz
/usr/src/custom-initramfs-5.10.61-gentoo.cpio.gz: gzip compressed data, max compression, from Unix, original size modulo 2^32 10344448
```

> I think that's all, the reason why it didn't work yesterday is because I am using a GPT disklabel

Using `GPT` as disk label only makes sense, if you are using a `uEFI system`. Moreover it forces you to format the `boot` partition as `FAT32`. Using `MSDOS` for `BIOS systems` gives you the freedom to choose between `FAT32` and `ext2`[1].

> I'm currently trying to transform a Pi that runs 32-bit RaspberryPi OS Light into one that runs 64-bit DietPi.

This feels like trying to run a 64-bit system with a lot of extra steps, while upgrading the stock image to a complete different system at the same time. This calls for issues. :sweat_smile:

My approach would be:

1. Install `64-bit DietPi` on a different system
2. Manually migrate every settings of the `32-bit Raspberry Pi OS` to the `64-bit DietPi system`

This is how you would do it on company level.

[1] Somewhere in-between it is explained in detail: https://wiki.gentoo.org/wiki/Handbook:AMD64/Full/Installation#Default_partitioning_scheme

**Comment by xanoni** [comment]
> Check, if the output of the `file` command looks something like this:

Ohhh... after the successful 64-bit upgrade, I was then able to use `pigz` and my initramfs is now standard ... so can no longer check.

> Using `GPT` as disk label only makes sense, if you are using a `uEFI system`. Moreover it forces you to format the `boot` partition as `FAT32`. Using `MSDOS` for `BIOS systems` gives you the freedom to choose between `FAT32` and `ext2`[1].

Interesting, I was not aware it could boot from `ext2`. However, I assume the advantages are negligible (for a 200MiB /boot partition) and that it's more of an ideological thing? :) 

> This feels like trying to run a 64-bit system with a lot of extra steps, while upgrading the stock image to a complete different system at the same time. This calls for issues. 😅

It's actually quite smooth ... didn't expect that ... has been running nicely so far. Also, the disk encryption performance went up as I probably mentioned somewhere else already... I was concerned it would go down.
